app/src/main/java/de/westnordost/streetcomplete/map/MainFragment.kt

```diff
@@ -396,16 +397,24 @@ class MainFragment : Fragment(R.layout.fragment_map_with_controls),
     }
 
     private fun onClickCompassButton() {
-        val mapFragment = mapFragment ?: return
-        val isNorthUp = mapFragment.cameraPosition?.rotation == 0f
-        if (!isNorthUp) {
-            mapFragment.updateCameraPosition(300) {
-                rotation = 0f
-                tilt = 0f
+        mapFragment?.run {
+            val isNorthUp = cameraPosition?.rotation == 0f
+            val isFlat = cameraPosition?.tilt == 0f
+            if (!isNorthUp) {
+                updateCameraPosition(300) {
+                    rotation = 0f
+                    tilt = 0f
+                }
+            }
+            if (isFollowingPosition) {
+                setIsCompassMode(!isCompassMode)
+            } else {
+                if (isNorthUp) {
+                    updateCameraPosition(300) {
+                        tilt = if (isFlat) PI.toFloat() / 5f else 0f
+                    }
+                }
             }
-        }
-        if (mapFragment.isFollowingPosition) {
-            setIsCompassMode(!mapFragment.isCompassMode)
         }
```

Comment:
Ah, didn't know `run`. So, that's another option to avoid the pesky "smart cast to X not possible because it is a field and might changed in the meantime":

1. `val x = x ?: return`
2. `x?.let { x -> ... }`
3. `val x = x; if (x != null) { ... }`
4. `x?.run { ... }`

I'd rather not use it, because
- consistency: not used anywhere else in the app
- I (personally) like "flat" functions, i.e. that don't have such a deep `{ { { ... } } }` hierarchy. I find it easier to read. 
- I tend to avoid functions that change what `this` is in a block is. Had some nasty bugs in the past that were not obvious when reading the code. (Shadowing properties of `this` of a containing block)

Reply:
Reverted to the old style. Nicer diff, anyway :P